lib/jsdom/living/nodes/HTMLInputElement-impl.js

```diff
@@ -590,6 +591,153 @@ class HTMLInputElementImpl extends HTMLElementImpl {
+            if (this.type === "checkbox" || this.type === "radio") {
+              return !this.checked;
+            } else {
+              return this.value === "";
+            }
+          }
+          return false;
+        },
+        // Gecko never allows the value to exceed the maxLength and minLength values,
+        // it would be simpler to do something like that here as well
+        tooLong: () => false,
+        tooShort: () => false,
+        // Is a Boolean indicating the value is greater than the maximum specified by the max attribute.
+        // The max attribute works with the following types:  number, range, date, datetime,
+        // datetime-local, month, time and week.
+        rangeOverflow: () => supportsRange() && this._maximum !== null && value() > this._maximum,
+        // Is a Boolean indicating the value is less than the minimum specified by the min attribute.
+        // The min attribute works with the following types:  number, range, date, datetime,
+        // datetime-local, month, time and week.
+        rangeUnderflow: () => supportsRange() && this._minimum !== null && value() < this._minimum,
+        // Is a Boolean indicating the value does not match the specified pattern.
+        patternMismatch: () => this.hasAttribute("pattern") &&
+            value() !== "" && !new RegExp(this.getAttribute("pattern")).test(value()),
+        // Is a Boolean indicating the value does not fit the rules determined by the step
+        // attribute(that is, it's not evenly divisible by the step value).
+        stepMismatch: () => {
+          if (supportsRange() && this.hasAttribute("step")) {
+            const number = value();
+            if (typeof number !== "number" || this.value === "") {
+              return false;
+            }
+            const step = parseFloatingPointNumber(this.getAttribute("step"));
+            if (!isNaN(step)) {
+              if (this.type === "week") {
+                return (parseWeekString(this.value).week - 1) % step !== 0;
+              } else if (this.type === "month") {
+                return (parseMonthString(this.value).month - 1) % step !== 0;
+              } else if (this.type === "date") {
+                return (parseDateString(this.value).day - 1) % step !== 0;
+              } else if (this.type === "datetime-local" || this.type === "datetime" ||
+                this.type == "time") {
+                return number % (step * 1000) !== 0;
+              }
+              return number % step !== 0;
+            }
+          }
+          return false;
+        },
+        // Is a Boolean indicating the value is not in the required syntax (when type is email or url).
+        typeMismatch: () => {
+          if (this.value === "") {
+            return false;
+          }
+          if (this.type === "email") {
+            const email = this.value;
+            const emailRegex = new RegExp("[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)" +
```

Comment:
The specification [suggests](https://html.spec.whatwg.org/multipage/input.html#valid-e-mail-address) a regular expression which differs a bit from this one. Probably safer to go with that instead.